include/picongpu/simulation/stage/CurrentInterpolationAndAdditionToEMF.hpp

```diff
@@ -0,0 +1,108 @@
+/* Copyright 2013-2019 Axel Huebl, Felix Schmitt, Heiko Burau, Rene Widera,
+ *                     Richard Pausch, Alexander Debus, Marco Garten,
+ *                     Benjamin Worpitz, Alexander Grund, Sergei Bastrakov
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/fields/currentInterpolation/CurrentInterpolation.hpp"
+#include "picongpu/fields/FieldJ.hpp"
+#include "picongpu/fields/maxwellSolver/Solvers.hpp"
+#include "picongpu/traits/GetMargin.hpp"
+
+#include <pmacc/dataManagement/DataConnector.hpp>
+#include <pmacc/dimensions/DataSpace.hpp>
+#include <pmacc/Environment.hpp>
+#include <pmacc/particles/traits/FilterByFlag.hpp>
+#include <pmacc/type/Area.hpp>
+
+#include <boost/mpl/count.hpp>
+
+#include <cstdint>
+
+
+namespace picongpu
+{
+namespace simulation
+{
+namespace stage
+{
+
+    /** Functor for the stage of the PIC loop performing current interpolation
+     *  and addition to grid values of the electromagnetic field
+     */
+    struct CurrentInterpolationAndAdditionToEMF
```

Comment:
Note: this is long on purpose, as this operation did and now does two things that are not really tightly coupled, probably to be separated some time later.